src/Psy/ExecutionLoop/Loop.php

```diff
@@ -27,17 +30,55 @@ class Loop
      */
     public function __construct(Configuration $config)
     {
-        // don't need this
+        $this->cleaner = $config->getCodeCleaner();
     }
 
     /**
-     * Run the execution loop.
-     *
-     * @param Shell $shell
+     * @return callable
      */
-    public function run(Shell $shell)
+    protected function getExecutionClosure()
     {
-        $loop = function ($__psysh__) {
+        return function (Shell &$__psysh__, $code = null) {
+
+            try {
+                extract($__psysh__->getScopeVariables());
+
+                // evaluate the current code buffer
+                ob_start();
+
+                set_error_handler(array($__psysh__, 'handleError'));
+                $_ = eval(!is_null($code) ? $code : $__psysh__->flushCode());
+                restore_error_handler();
+
+                $__psysh_out__ = ob_get_contents();
+                ob_end_clean();
+
+                $__psysh__->writeStdout($__psysh_out__);
+                $__psysh__->writeReturnValue($_);
+
+                // a bit of housekeeping
+                unset($__psysh_out__);
+                unset($code);
+
+                $__psysh__->setScopeVariables(get_defined_vars());
+            } catch (BreakException $_e) {
+                throw $_e;
+            } catch (\Exception $_e) {
+                restore_error_handler();
+                if (ob_get_level() > 0) {
+                    ob_end_clean();
+                }
+                $__psysh__->writeException($_e);
+            }
+        };
+    }
+
+    /**
+     * @return callable
+     */
+    protected function getLoopClosure(\Closure $executionClosure)
+    {
+        return function (Shell &$__psysh__) use ($executionClosure) {
```

Comment:
This will put `$executionClosure` in the local scope while running the REPL. I've been very careful to _only_ expose `$__psysh__` so that the amount of unexpected PsySH stuff "leaking" into the REPL is very limited. And then I added a CodeCleaner pass to make it hard for you to even touch that one variable :P

Is there a way to do this without exposing another variable name?